File: apps/api/src/auth/auth.controller.ts
import { Body, Controller, Get, Post, Req, Res } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { ApiTags } from '@nestjs/swagger';
import type { Request, Response } from 'express';
import {
  type AuthResult,
  EmployeeOtpRequestSchema,
  type EmployeeOtpRequestInput,
  EmployeeOtpVerifySchema,
  type EmployeeOtpVerifyInput,
  type OtpRequestResult,
  type Session,
  StaffLoginSchema,
  type StaffLoginInput,
} from '@ihrms/shared';
import type { Env } from '../config/env.validation';
import { ZodValidationPipe } from '../common/pipes/zod-validation.pipe';
import type { AuditActor } from '../audit/audit.service';
import { AuthService, type IssuedSession } from './auth.service';
import { refreshCookieOptions } from './cookies';
import { Public } from './decorators/public.decorator';
import { CurrentUser } from './decorators/current-user.decorator';
import { type Principal, toSession } from './principal';

interface AuthRequest extends Request {
  auditActor?: AuditActor;
}

@ApiTags('auth')
@Controller('auth')
export class AuthController {
  constructor(
    private readonly auth: AuthService,
    private readonly config: ConfigService<Env, true>,
  ) {}

  @Public()
  @Post('login')
  login(
    @Body(new ZodValidationPipe(StaffLoginSchema)) body: StaffLoginInput,
    @Req() req: AuthRequest,
    @Res({ passthrough: true }) res: Response,
  ): Promise<AuthResult> {
    return this.auth.loginStaff(body).then((issued) => this.complete(issued, req, res));
  }

  @Public()
  @Post('employee/request-otp')
  requestOtp(
    @Body(new ZodValidationPipe(EmployeeOtpRequestSchema)) body: EmployeeOtpRequestInput,
  ): Promise<OtpRequestResult> {
    return this.auth.requestEmployeeOtp(body);
  }

  @Public()
  @Post('employee/verify-otp')
  verifyOtp(
    @Body(new ZodValidationPipe(EmployeeOtpVerifySchema)) body: EmployeeOtpVerifyInput,
    @Req() req: AuthRequest,
    @Res({ passthrough: true }) res: Response,
  ): Promise<AuthResult> {
    return this.auth.verifyEmployeeOtp(body).then((issued) => this.complete(issued, req, res));
  }

  @Public()
  @Post('refresh')
  refresh(@Req() req: AuthRequest, @Res({ passthrough: true }) res: Response): Promise<AuthResult> {
    const cookieName = this.config.get('REFRESH_COOKIE_NAME', { infer: true });
    const token = req.cookies?.[cookieName];
    return this.auth.refresh(token).then((issued) => this.complete(issued, req, res));
  }

  @Public()
  @Post('logout')
  logout(@Res({ passthrough: true }) res: Response): { ok: true } {
    const cookieName = this.config.get('REFRESH_COOKIE_NAME', { infer: true });
    res.clearCookie(cookieName, refreshCookieOptions(this.config));
    return { ok: true };
  }

  @Get('me')
  me(@CurrentUser() principal: Principal): Session {
    return toSession(principal);
  }

  /** Set the refresh cookie and the audit actor (req.user is unset on @Public routes). */
  private complete(issued: IssuedSession, req: AuthRequest, res: Response): AuthResult {
    const cookieName = this.config.get('REFRESH_COOKIE_NAME', { infer: true });
    res.cookie(cookieName, issued.refreshToken, refreshCookieOptions(this.config));

    const s = issued.result.session;
    req.auditActor =
      s.type === 'USER'
        ? { actorType: 'USER', actorId: s.userId, companyId: s.companyId }
        : { actorType: 'EMPLOYEE', actorId: s.employeeId, companyId: s.companyId };

    return issued.result;
  }
}

File: apps/api/src/main/java/com/ihrms/auth/AuthController.java
package com.ihrms.auth;

import com.ihrms.audit.AuditActor;
import com.ihrms.audit.AuditInterceptor;
import com.ihrms.auth.AuthService.IssuedSession;
import com.ihrms.auth.dto.AuthDtos.AuthResult;
import com.ihrms.auth.dto.AuthDtos.ChangePasswordRequest;
import com.ihrms.auth.dto.AuthDtos.OkResponse;
import com.ihrms.auth.dto.AuthDtos.OtpRequest;
import com.ihrms.auth.dto.AuthDtos.OtpRequestResult;
import com.ihrms.auth.dto.AuthDtos.OtpVerifyRequest;
import com.ihrms.auth.dto.AuthDtos.StaffLoginRequest;
import com.ihrms.auth.dto.SessionView;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.server.ResponseStatusException;

/**
 * Auth endpoints (contract §3.2). POST -> 201, GET -> 200. The refresh token is delivered
 * only in the httpOnly {@code ihrms_refresh} cookie; the access token is in the body. All
 * but {@code /auth/me} are public (see SecurityConfig); /auth/me requires authentication.
 */
@RestController
@RequestMapping("/auth")
public class AuthController {

  private final AuthService auth;
  private final RefreshCookies cookies;

  public AuthController(AuthService auth, RefreshCookies cookies) {
    this.auth = auth;
    this.cookies = cookies;
  }

  /** Staff sign-in: email + password -> session (resolves a User only; generic denial otherwise). */
  @PostMapping("/login")
  @ResponseStatus(HttpStatus.CREATED)
  public AuthResult login(
      @Valid @RequestBody StaffLoginRequest body,
      HttpServletRequest request,
      HttpServletResponse response) {
    return complete(auth.loginStaff(body), request, response);
  }

  /** Staff self-service password change (authenticated). Employees have no password. */
  @PostMapping("/change-password")
  @ResponseStatus(HttpStatus.CREATED)
  public OkResponse changePassword(
      @Valid @RequestBody ChangePasswordRequest body,
      @AuthenticationPrincipal IhrmsPrincipal principal,
      HttpServletRequest request) {
    if (!(principal instanceof IhrmsPrincipal.User user)) {
      throw new ResponseStatusException(HttpStatus.FORBIDDEN, "Only staff accounts have a password");
    }
    auth.changePassword(user, body, request.getRemoteAddr());
    return new OkResponse(true);
  }

  /** Employee sign-in start: full name + email -> OTP. Enumeration-safe generic result. */
  @PostMapping("/request-otp")
  @ResponseStatus(HttpStatus.CREATED)
  public OtpRequestResult requestOtp(@Valid @RequestBody OtpRequest body) {
    return auth.requestOtp(body);
  }

  /** Employee verify: email + code -> session (principal type + role + scope from the account). */
  @PostMapping("/verify-otp")
  @ResponseStatus(HttpStatus.CREATED)
  public AuthResult verifyOtp(
      @Valid @RequestBody OtpVerifyRequest body,
      HttpServletRequest request,
      HttpServletResponse response) {
    return complete(auth.verifyOtp(body), request, response);
  }

  @PostMapping("/refresh")
  @ResponseStatus(HttpStatus.CREATED)
  public AuthResult refresh(
      @CookieValue(value = "${app.refresh-cookie-name:ihrms_refresh}", required = false)
          String refreshToken,
      HttpServletRequest request,
      HttpServletResponse response) {
    return complete(auth.refresh(refreshToken), request, response);
  }

  @PostMapping("/logout")
  @ResponseStatus(HttpStatus.CREATED)
  public OkResponse logout(HttpServletResponse response) {
    response.addHeader(HttpHeaders.SET_COOKIE, cookies.clear().toString());
    return new OkResponse(true);
  }

  @GetMapping("/me")
  public SessionView me(@AuthenticationPrincipal IhrmsPrincipal principal) {
    return auth.sessionFor(principal);
  }

  /** Set the refresh cookie and the audit actor (no SecurityContext on @Public auth routes). */
  private AuthResult complete(
      IssuedSession issued, HttpServletRequest request, HttpServletResponse response) {
    response.addHeader(HttpHeaders.SET_COOKIE, cookies.set(issued.refreshToken()).toString());
    request.setAttribute(
        AuditInterceptor.AUDIT_ACTOR_ATTR, AuditActor.from(issued.result().session()));
    return issued.result();
  }
}
